### PR TITLE
DM-12371: Implement minimal butler Registry prototype with sqlalchemy core

### PR DESCRIPTION
**Pull request by pschella**

### CODE REVIEW

File: python/lsst/butler/datasets.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# 
+# LSST Data Management System
+#
+# Copyright 2008-2017  AURA/LSST.
```
Comment (timj) [inline python/lsst/butler/datasets.py]: Surely not 2008?

File: python/lsst/butler/datasets.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+
+from types import MappingProxyType
+from .utils import slotValuesAreEqual, slotValuesToHash
+from .storageClass import StorageClass
+from .units import DataUnitTypeSet
+
+
+class DatasetType:
```
Comment (timj) [inline python/lsst/butler/datasets.py]: I know it's a prototype but the complete lack of doc strings makes it hard to work out what the attributes in the classes are meant to be.

File: python/lsst/butler/datasets.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        return self._name
+
+    @property
+    def units(self):
+        return self._units
+
+
+class DatasetRef(DatasetLabel):
+
+    __slots__ = ("_type", "_units", "_producer",
```
Comment (timj) [inline python/lsst/butler/datasets.py]: I was under the impression that subclasses with `__slots__` couldn't reuse the same attributes. `_units` is already defined in the base class.

File: python/lsst/butler/datasets.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    def actualConsumers(self):
+        return MappingProxyType(self._actualConsumers)
+
+    def makePath(self, run, template=None):
+        raise NotImplementedError("TODO")
+
+
+class DatasetHandle(DatasetRef):
+
+    __slots__ = "_datasetId", "_registryId", "_uri", "_components", "_run"
```
Comment (timj) [inline python/lsst/butler/datasets.py]: you are inconsistent with defining these. Sometimes you use parentheses and sometimes you don't.

File: python/lsst/butler/quantum.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+from .datasets import DatasetRef
+from .run import Run
+
+
+class Quantum:
+
+    __slots__ = ("_quantumId", "_registryId", "_run", "_task", "_predictedInputs", "_actualInputs")
+
+    _currentId = -1
```
Comment (timj) [inline python/lsst/butler/quantum.py]: This starts at -1 but the one above started at 0.

File: python/lsst/butler/quantum.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    @classmethod
+    def getNewId(cls):
+        cls._currentId += 1
+        return cls._currentId
+
+    def __init__(self, run, task=None, quantumId=None, registryId=None):
+        assert isinstance(run, Run)
+        self._quantumId = quantumId
```
Comment (timj) [inline python/lsst/butler/quantum.py]: This is using an ID but nothing is calling `getNewId`.

File: python/lsst/butler/quantum.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        return self._predictedInputs
+
+    @property
+    def actualInputs(self):
+        return self._actualInputs
+
+    def addPredictedInput(self, ref):
+        assert isinstance(ref, DatasetRef)
+
+        if ref.type.name not in self._actualInputs:
```
Comment (timj) [inline python/lsst/butler/quantum.py]: Without any comments or documentation in the classes it's a bit hard to work out why the check is for `ref.type.name` but the key used is then `ref.name`.

File: python/lsst/butler/registry.py [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+from .datasets import DatasetType, DatasetHandle, DatasetRef, DatasetLabel
+from .run import Run
+from .quantum import Quantum
+from .units import DataUnit, DataUnitTypeSet
+from .storageClass import StorageClass
+
+
+class Registry:
+    """Basic SQL backed registry.
+    """
```
Comment (timj) [inline python/lsst/butler/registry.py]: Attribute documentation is missing.

File: python/lsst/butler/registry.py [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+
+        The :ref:`Quantum` that generated the :ref:`Dataset` can optionally be provided to add provenance information.
+
+        :param ref: a :ref:`DatasetRef` that identifies the :ref:`Dataset` and contains its :ref:`DatasetType`.
+
+        :param str uri: the :ref:`URI` that has been associated with the :ref:`Dataset` by a :ref:`Datastore`.
+
+        :param dict components: if the :ref:`Dataset` is a composite, a ``{name : URI}`` dictionary of its named components and storage locations.
+
+        :param Run run: the :ref:`Run` instance that produced the Dataset.  Ignored if ``producer`` is passed (:py:attr:`producer.run <Quantum.run>` is then used instead).  A Run must be provided by one of the two arguments.
```
Comment (timj) [inline python/lsst/butler/registry.py]: docstring lines are too long. stick to 79 characters.

File: python/lsst/butler/units.py [hunk 10]
```diff
@@ -0,0 +1,992 @@
...
+        raise NotImplementedError("pure virtual")
+
+    def pkey(self):
+        return NotImplemented("pure virtual")
+
+    def invariantHash(self):
+        """
+        Compute a hash that is invariant accross Python sessions, and hence can be stored in a database.
+        """
+        return sha512(b''.join(str(v).encode() for v in self.pkey)).digest()
```
Comment (timj) [inline python/lsst/butler/units.py]: If you want to be invariant I suggest you specify the encoding rather than relying on the default.

File: tests/SConscript [hunk 11]
```diff
@@ -0,0 +1,3 @@
+# -*- python -*-
+from lsst.sconsUtils import scripts
+scripts.BasicSConscript.tests()
```
Comment (timj) [inline tests/SConscript]: Add `pyList=[]` as argument for `tests()`. Also add a `setup.cfg` to enable flake8 testing.

**Review by timj — COMMENTED** [review]
I've had a quick scan of all of it. I haven't really grokked it all yet but it looks reasonable. In particular I've never really used sqlalchemy before so that's a bit hard to follow.

I think I'd like LSST concepts and CCD sensor concepts in their own files to make it clear to external adopters of butler which files contain LSST oddness and which are generic.

File: python/lsst/butler/datasets.py (hunk 3 above)
Comment (pschella) [reply python/lsst/butler/datasets.py]: It can, but it wastes memory. So I have removed it.

File: python/lsst/butler/datasets.py (hunk 4 above)
Comment (pschella) [reply python/lsst/butler/datasets.py]: That is probably because I wrote some of it, and Jim some other parts ;) I'm now using parentheses everywhere.

File: python/lsst/butler/quantum.py (hunk 5 above)
Comment (pschella) [reply python/lsst/butler/quantum.py]: Both are pre-increment, so should start at `-1`. Note that the whole ID thing is ugly, flaky and needs to be fixed. The only reason for it existing is because `sqlite3` doesn't allow `autoincrement` compound-primary-keys. I should probably create a ticket for this...

Comment (pschella) [reply python/lsst/butler/quantum.py]: DM-13063 created.

File: python/lsst/butler/quantum.py (hunk 6 above)
Comment (pschella) [reply python/lsst/butler/quantum.py]: That is intentional (if ugly). Will be fixed on DM-13063.

File: python/lsst/butler/quantum.py (hunk 7 above)
Comment (pschella) [reply python/lsst/butler/quantum.py]: It should have been `ref.type.name`. Updated and added docstrings for `Quantum`.

File: python/lsst/butler/registry.py (hunk 9 above)
Comment (pschella) [reply python/lsst/butler/registry.py]: Also converted to numpydoc.
